### ISSUE TITLE
Consider removing "Prefix Unexported Globals with _" recommendation

### ISSUE THREAD

**Issue by shyiko** [uber-go/guide#26]
I challenge the rational behind this recommendation. 

_ prefix:
- makes naming rules confusing & opens up a possibility for an abuse. 
My observation is that people start to use _ prefix left and right
- is redundant & can always be avoided, e.g. [Prefix Unexported Globals with _](https://github.com/uber-go/guide/blob/master/style.md#prefix-unexported-globals-with-_) example can be replaced with 

```go
// foo.go

const (
  defaultFooPort = 8080
  defaultFooUser = "user"
)

// bar.go

func Bar() {
  defaultBarPort := 9090
  ...
}
```

or simply

```go
// foo.go

const (
  defaultPort = 8080
  defaultUser = "user"
)

// bar.go

func Bar() {
  port := 9090
  ...
}
```

- package-level variables can be shown as such/highlighted by the IDE if that's your preference (e.g. in GoLand, go to Settings -> Editor -> Code Scheme -> Go -> Variable ->  Package local variable). 
- feels like python-ism / Hungarian notion in disguise
- generally inconsistent with the rest of Go community

Based on the above I believe prefixing variables with _ should be considered not just unnecessary but harmful.

**Comment by prashantv** [comment]
> makes naming rules confusing & opens up a possibility for an abuse.

The existing naming rules are:
 - capitalize if exported
 - lowercase if unexported.

We're adding to the 2nd rule, `_` prefix if global. Not sure where the confusion stems from, but "unexported global" is pretty specific. I haven't seen any abuse or misuse of this and I've reviewed a lot of Go code at Uber. If there is misuse, given how specific the recommendation is, we should be able to converge to doing it correctly.

> is redundant & can always be avoided, e.g. Prefix Unexported Globals with _ example can be replaced with [...]

It can be avoided with renaming, and it's easy to do in the case of `defaultPort`, but the issues are:
 * The global is often named something like "port"
 * The user using a shadowed variable "port" may not realize they're shadowing.

I don't think it's redundant, the whole point is to avoid accidentally referencing the variable in the wrong scope, it's easy to not realize a global is shadowed since global scoped variables are declared farther away (E.g., in any location in any of the files in the package).

> package-level variables can be shown as such/highlighted by the IDE if that's your preference (e.g. in GoLand, go to Settings -> Editor -> Code Scheme -> Go -> Variable -> Package local variable).

IDE-specific solutions don't help all users, or in code reviews.

> feels like python-ism / Hungarian notion in disguise

Go already uses capitalization for exported/unexported. The underscore is in the same vein, although it does introduce a character that's more common in other languages.

> generally inconsistent with the rest of Go community

I agree that this is different from the Go community, but since it's localized (affects just the internals of a package), and it's not something that's explicitly discouraged, I don't think it's a big concern.

-1 on removing this recommendation, I think it helps with readability and code review, and I don't agree with most of the concerns.

**Comment by shyiko** [comment]
The only thing I would ask is that before dismissing the request please consider the following:

> makes naming rules confusing & opens up a possibility for an abuse

`ripgrep`'ing over monorepo:

```bash
# OK: count global var _example...
$ rg "^var _[a-z]" | wc -l
798

# not OK: count local var _example... (example of misuse)
$ rg "\t+var _[a-z]" | grep -v -F 'interface{}' | wc -l
25
# not OK: count global var lowercase... without _ prefix
$ rg "^var [^_(A-Z]" | wc -l
2147
# not OK: could global var _err...
$ rg "^var _err" | wc -l
4
```

> IDE-specific solutions don't help all users, or in code reviews.

I agree. That comment about Goland was intended primarily for the people that might not be aware of existing support. It's certainly not something that can help with the code reviews or people using vim for go development. BUT. I'd argue that things like these should be automated (and not rely on code review process) so that they could be caught even before the diffs are submitted (this saves time on code reviews & keeps violations from slipping through the cracks).  
`go vet -vettool=$(which shadow) ./...` is pretty good at reporting shadow declarations (`shadow` operates on file not package level but that shouldn't be too hard to fix; I'd be even willing to do this).  

> the whole point is to avoid accidentally referencing the variable in the wrong scope, it's easy to not realize a global is shadowed since global scoped variables are declared farther away

Fair enough. It has never been the problem for me personally but I can see how this could be problematic if the number of files in the package grows out of control. Keeping packages small should help but I do agree with the point. 

> I agree that this is different from the Go community, but since it's localized (affects just the internals of a package), and it's not something that's explicitly discouraged, I don't think it's a big concern.

My main concern is an additional context switch. All other recommendations in this style guide (at least in my opinion) are more or less universally true. This one, on the other hand, forces a different code style in & out of Uber. This affects both internal (employees who need to change the way they write code during the day and in the night :)) & external users (contributors to Uber's OSS projects who may be familiar with this style guide).

**tldr;** I propose to remove this rule & use go vet + shadow for reporting shadow declarations.

**Comment by shyiko** [comment]
Another reason for going with static code analysis is code like this:

```go
// a.go
func f() {
}

// b.go
func f2() {
	var f = func() {}
	...
	f()
	// Removing `var f = func() {}` will still compile.
}
```
Same kind of trap and yet following [Prefix Unexported Globals with _](https://github.com/uber-go/guide/blob/master/style.md#prefix-unexported-globals-with-_) does not help (I understand this case is less likely to happen but it just serves as an example of when prefixing _ fails and static analysis still works).

**Comment by prashantv** [comment]
> ripgrep'ing over 
 (approximation) [...]
> My observation is that people start to use _ prefix left and right

Your own analysis shows that misuse is pretty limited: <30 instances is pretty minor, and definitely not frequent misuse as the earlier comment tried to suggest.

I agree that users aren't doing it consistently, but I wouldn't count that as misuse.

> go vet -vettool=$(which shadow) ./... is pretty good at reporting shadow declarations (shadow operates on file not package level but that shouldn't be too hard to fix; I'd be even willing to do this).

If shadow works on the package level, and we don't have false positives, then I think we should consider that as an option.

> My main concern is an additional context switch. All other recommendations in this style guide (at least in my opinion) are more or less universally true. This one, on the other hand, forces a different code style in & out of Uber. This affects both internal (employees who need to change the way they write code during the day and in the night :)) & external users (contributors to Uber's OSS projects who may be unfamiliar with this style guide).

Understand the concern, but the focus of this style guide is to help with Uber's code, and this recommendation's goal is to improve correctness. The cost is minimal (and `_` prefix), and has benefits that we've run into while writing Go code, hence the recommendation.

> Another reason for going with static code analysis is code like this:
> 	var f = func() {}

Agree, this isn't perfect, but just because it doesn't catch a rare edge case, I don't think that's a good reason to abandon this recommendation. This is trying to reduce the likelihood of such issues, not eliminate the issue completely.

---

If we find that there's a tooling solution that works well, and we can enable it easily in all our code (internal, and external GitHub repos), then I'm open to reevaluating this recommendation. At this point, it seems like the tooling doesn't even support the alternative.

**Comment by mway** [comment]
Hey @shyiko, thanks for the feedback and discussion. As discussed here, let's revisit when Go tooling catches up and can make this guidance obsolete. In the meantime, the current guidance will stand.

Please reach out if you have other ideas and/or concerns about guidance!
